### DISCUSSION TITLE
MacOS 15 Context3D not available

### DISCUSSION THREAD

**Discussion by hardcoremore** [airsdk/Adobe-Runtime-Support#3596] [General]
I have just bought Mac Mini M4 and have setup my project with IntelliJ. I am using latest AIR SDK 51.1.3, Starling 2.2 and SWF version 51. Everything seems fine except that when I run application I got "Context3D is not available" error.

It looks like adl is not allowed access to GPU. Did anyone have this issue before. Is there something I can do?

This is my app.xml config

<fullScreen>false</fullScreen>

        
        <renderMode>direct</renderMode>

        
        <depthAndStencil>true</depthAndStencil>

		
		

		
		

		<aspectRatio>landscape</aspectRatio>

		<autoOrients>false</autoOrients>

        <visible>true</visible>

**Comment by ajwfrost** [comment]
Just out of interest, does it change things if you try with the `depthAndStencil` value set to false?
And do you see any output in the "Console" application, if you open that and start streaming events, and filter for your application name as the process?

thanks

**Comment by hardcoremore** [comment]
Disabling stencil does not help. This is from IntelliJ debug console:


```
"/Applications/IntelliJ IDEA.app/Contents/jbr/Contents/Home/bin/java" -Dapplication.home=/Users/hardcoremore/AdobeAIR/sdk/AIRSDK_MacOS-51.1.3.1 -Dfile.encoding=UTF-8 -Djava.awt.headless=true -Duser.language=en -Duser.region=en -Xmx512m -classpath /Users/hardcoremore/AdobeAIR/sdk/AIRSDK_MacOS-51.1.3.1/lib/legacy/fdb.jar flex.tools.debugger.cli.DebugCLI
Adobe fdb (Flash Player Debugger) [build development]
Copyright (c) 2004-2007 Adobe, Inc. All rights reserved.
Waiting for Player to connect
/Users/hardcoremore/AdobeAIR/sdk/AIRSDK_MacOS-51.1.3.1/bin/adl -profile extendedDesktop /Users/hardcoremore/AdobeAIR/Adobe_Flash_Builder_Work_Space/ShapesInMotionMAC/out/production/ShapesInMotionMAC/ShapesInMotionMAC-descriptor.xml /Users/hardcoremore/AdobeAIR/Adobe_Flash_Builder_Work_Space/ShapesInMotionMAC/out/production/ShapesInMotionMAC
Player connected; session starting.
[trace] LEVEL COUNT:  51
[trace] [Starling] Warning: 'enableErrorChecking' has a negative impact on performance. Never activate for release builds!
[trace] [Starling] Context3D not available! Possible reasons: wrong renderMode or missing device support.
Player session terminated
```

**Reply by ajwfrost** [reply]
Okay thanks. We may need to send a version with a bit more debugging that should come out of it; I actually meant though, the macOS "Console" app - i.e. hit Command-Space and type "Console" and you'll get the native macOS console which gives out a lot of information for debugging/troubleshooting.. (but, you have to filter it, otherwise there's a lot of irrelevance!)

thanks

**Comment by hardcoremore** [comment]
This is all I can find for adl proces inside launched.log

2024-12-09 18:35:20.994868 (pid/7684 [adl]) <Notice>: uncorking exec source upfront
2024-12-09 18:35:20.994879 (pid/7684 [adl]) <Notice>: created
2024-12-09 18:35:20.995259 (pid/7684 [adl]) <Notice>: uncorking domain

**Reply by ajwfrost** [reply]
Thanks; I'll try to send you a build of the runtime together that displays a bit more information..!

**Reply by hardcoremore** [reply]
Awesome. Thanks so much!

**Reply by hardcoremore** [reply]
Hi Andy,

I have resolved this. It was because I did not set Custom Template to use my Application Descriptor inside IntelliJ. It was using generated one which probably didn't have set option for rendering mode.   

It may be useful to add some log (for example logging current requested render mode inside app descriptor) when error Context.3D is thrown. Because I assume that generated descriptor  has renderMode as CPU.


Thanks!

**Reply by ajwfrost** [reply]
Well discovered!

Currently the error message for this would be:
```
Error #3702: Context3D not available.
```
There is actually an internal error reason so we could adjust the error string there, whilst keeping the code the same. So e.g one of:
```
Error #3702: Context3D not available. Reason 0: Invalid renderMode value.
Error #3702: Context3D not available. Reason 1: Software mode requested.
Error #3702: Context3D not available. Reason 2: User has disabled hardware acceleration.
Error #3702: Context3D not available. Reason 3: Unsupported graphics driver.
Error #3702: Context3D not available. Reason 4: Graphics hardware unavailable.
```

Then it might also be beneficial to update Starling to also give this information out, currently it just checks the code:
https://github.com/Gamua/Starling-Framework/blob/master/starling/src/starling/core/Starling.as#L608

thanks

**Reply by hardcoremore** [reply]
Yes this would be awesome. It would be great to have additional info in the error message!
